### PR TITLE
feat: add comprehensive image extraction support

### PR DESCRIPTION
**Pull request by Goldziher**
## Summary

This PR adds comprehensive image extraction capabilities to Kreuzberg, enabling users to extract embedded images from various document formats and optionally run OCR on them.

### Core Features

- **Image Extraction**: Extract embedded images from PDF, PPTX, HTML, DOCX, and EML documents
- **OCR Processing**: Run OCR on extracted images using Tesseract, EasyOCR, or PaddleOCR
- **Image Filtering**: Configurable dimension limits and deduplication
- **Performance Optimization**: Batch processing optimizations and memory management
- **API Integration**: Full REST API support with query parameters and header configuration

### New Types and Configuration

- `ExtractedImage` dataclass with format, dimensions, page info, and binary data
- `ImageOCRResult` dataclass combining image and OCR results
- Enhanced `ExtractionConfig` with image extraction parameters
- Comprehensive benchmarking suite for performance validation

### Documentation

- Complete configuration examples for all scenarios
- API documentation with curl examples and response formats
- Extensive usage examples covering different document types
- Performance optimization guidelines
- API reference documentation for new types

### Test Coverage

- 35+ new test files covering image extraction workflows
- Integration tests for all supported document formats
- Error handling and edge case validation
- Memory limits and performance testing

### Performance

- Optimized batch processing with dynamic worker allocation
- 93.6% speed improvement (338.94s → 21.79s) over initial implementation
- 92.6% memory reduction (7GB → 520MB)
- Comprehensive benchmarking results included

## Test Plan

- [x] All existing tests pass
- [x] New image extraction tests pass
- [x] Integration tests with real documents
- [x] Performance benchmarks completed
- [x] Memory usage validation
- [x] API endpoint testing
- [x] Documentation examples verified

### CODE REVIEW

File: benchmarks/optimization_summary.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Kreuzberg Batch Processing Optimization Results
```
Comment (Goldziher) [inline benchmarks/optimization_summary.md:1]: Cleanup: remove this file

File: benchmarks/results/final_batch_validation.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (Goldziher) [inline benchmarks/results/final_batch_validation.json:1]: Cleanup: remove the results folder and git ignore it

File: benchmarks/src/profiler.py [hunk 3]
```diff
@@ -131,13 +125,11 @@ def profile_function(
...
 
-            process = subprocess.Popen(
-                cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
-            )
+            process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
 
             result = func()
 
-            stdout, stderr = process.communicate()
+            _stdout, _stderr = process.communicate()
```
Comment (Goldziher) [inline benchmarks/src/profiler.py:132]: Cleanup any unused variables

File: benchmarks/src/profiler.py [hunk 4]
```diff
@@ -146,7 +138,7 @@ def profile_function(
         finally:
             temp_script.unlink(missing_ok=True)
 
-    def _generate_function_call_code(self, func: Callable[[], T]) -> str:
+    def _generate_function_call_code(self, _func: Callable[[], T]) -> str:
```
Comment (Goldziher) [inline benchmarks/src/profiler.py:141]: Cleanup unused params

File: benchmarks/src/profiler.py [hunk 5]
```diff
@@ -158,7 +150,7 @@ def _generate_function_call_code(self, func: Callable[[], T]) -> str:
 
 @contextmanager
 def profile_benchmark(
-    flame_config: FlameGraphConfig | None = None,
+    _flame_config: FlameGraphConfig | None = None,
```
Comment (Goldziher) [inline benchmarks/src/profiler.py:153]: Systematically cleanup the benchmarks files for unused assignments, variables, parameters etc

File: docs/user-guide/extraction-configuration.md [hunk 6]
```diff
@@ -35,6 +35,14 @@ auto_detect_document_type = true
 document_classification_mode = "text"  # or "vision"
 type_confidence_threshold = 0.5
 
+# Image extraction configuration
+extract_images = true              # Extract embedded images from documents
```
Comment (Goldziher) [inline docs/user-guide/extraction-configuration.md:39]: Clean this file up properly, lots of orphaned # symbols

File: kreuzberg/_api/main.py [hunk 7]
```diff
@@ -223,6 +223,13 @@ async def handle_files_upload(  # noqa: PLR0913
     ocr_backend: Literal["tesseract", "easyocr", "paddleocr"] | None = None,
     auto_detect_language: str | bool | None = None,
     pdf_password: str | None = None,
+    extract_images: str | bool | None = None,
+    ocr_extracted_images: str | bool | None = None,
+    image_ocr_backend: Literal["tesseract", "easyocr", "paddleocr"] | None = None,
```
Comment (Goldziher) [inline kreuzberg/_api/main.py:255]: Let's create an ImageOCRConfig dataclass instead of. Passing 5 or 6 new parameters to the config.

File: kreuzberg/_api/main.py [hunk 8]
```diff
@@ -316,18 +355,41 @@ async def get_configuration() -> ConfigurationResponse:
...
 
 def _pil_image_encoder(obj: Any) -> str:
-    """Convert PIL Image to base64 string for JSON serialization."""
     buffer = io.BytesIO()
     obj.save(buffer, format="PNG")
     img_str = base64.b64encode(buffer.getvalue()).decode()
     return f"data:image/png;base64,{img_str}"
 
 
+def _extracted_image_encoder(obj: ExtractedImage) -> dict[str, Any]:
```
Comment (Goldziher) [inline kreuzberg/_api/main.py]: Don't return dict, return a typed dict

File: kreuzberg/_api/main.py [hunk 9]
```diff
@@ -316,18 +355,41 @@ async def get_configuration() -> ConfigurationResponse:
...
+        "page_number": obj.page_number,
+        "dimensions": obj.dimensions,
+        "colorspace": obj.colorspace,
+        "bits_per_component": obj.bits_per_component,
+        "is_mask": obj.is_mask,
+        "description": obj.description,
+    }
+
+
+def _image_ocr_result_encoder(obj: ImageOCRResult) -> dict[str, Any]:
```
Comment (Goldziher) [inline kreuzberg/_api/main.py]: Same, should be typed dict

File: kreuzberg/_api/main.py (hunk 7 above)
Comment (Goldziher) [reply kreuzberg/_api/main.py:255]: We also have to support proper OCR configs for these, so this dataclass should extend or share a base class with the other OCR configs. This requires further planning

File: kreuzberg/_extractors/_base.py [hunk 10]
```diff
@@ -52,11 +75,193 @@ def _apply_quality_processing(self, result: ExtractionResult) -> ExtractionResul
...
+            entities=result.entities,
+            keywords=result.keywords,
             detected_languages=result.detected_languages,
-            tables=result.tables,
+            document_type=result.document_type,
+            document_type_confidence=result.document_type_confidence,
+            layout=result.layout,
         )
+
+    def _check_image_memory_limits(self, images: list[ExtractedImage]) -> list[ExtractedImage]:
```
Comment (Goldziher) [inline kreuzberg/_extractors/_base.py:98]: This method should be refactored

File: kreuzberg/_extractors/_base.py [hunk 11]
```diff
@@ -52,11 +75,193 @@ def _apply_quality_processing(self, result: ExtractionResult) -> ExtractionResul
...
+                    img.filename or "unknown",
+                    img_size,
+                    MAX_SINGLE_IMAGE_SIZE,
+                )
+            else:
+                filtered.append(img)
+
+        return filtered
+
+    def _deduplicate_images(self, images: list[ExtractedImage]) -> list[ExtractedImage]:
```
Comment (Goldziher) [inline kreuzberg/_extractors/_base.py:171]: This method should be refactored

File: kreuzberg/_extractors/_base.py [hunk 12]
```diff
@@ -52,11 +75,193 @@ def _apply_quality_processing(self, result: ExtractionResult) -> ExtractionResul
...
+        cfg = _config_for_backend()
+        backend = get_ocr_backend(backend_name)
+
+        results: list[ImageOCRResult] = []
+        min_w, min_h = self.config.image_ocr_min_dimensions
+        max_w, max_h = self.config.image_ocr_max_dimensions
+
+        tasks = []
+
+        async def _ocr_one(target: ExtractedImage) -> ImageOCRResult:
```
Comment (Goldziher) [inline kreuzberg/_extractors/_base.py]: Should single, not one, also This method should be refactored

File: kreuzberg/_extractors/_email.py [hunk 13]
```diff
@@ -111,20 +110,83 @@ def _extract_email_body(self, parsed_email: dict[str, Any], text_parts: list[str
...
+                text_parts.append(converted_text)
             else:
-                clean_html = _HTML_TAG_PATTERN.sub("", html_content)
+                cleaned = re.sub(r"<script[^>]*>.*?</script>", "", html_content, flags=re.IGNORECASE | re.DOTALL)
+                cleaned = re.sub(r"<style[^>]*>.*?</style>", "", cleaned, flags=re.IGNORECASE | re.DOTALL)
+                clean_html = _HTML_TAG_PATTERN.sub("", cleaned)
                 clean_html = unescape(clean_html)
-                text_parts.append(f"\n{clean_html}")
+                clean_html = (
+                    clean_html.replace("\u201c", '"')
```
Comment (Goldziher) [inline kreuzberg/_extractors/_email.py:115]: Create a precompiled regex constant at the top of the file and use it instead of these replaces

File: kreuzberg/_extractors/_email.py [hunk 14]
```diff
@@ -111,20 +110,83 @@ def _extract_email_body(self, parsed_email: dict[str, Any], text_parts: list[str
...
+            if isinstance(att, dict):
+                n = att.get("name")
+                if isinstance(n, str) and n:
+                    name_val = n
+            names.append(name_val)
+        metadata["attachments"] = names
+        if names:
+            text_parts.append("Attachments: " + ", ".join(names))
+
+    def _extract_images_from_attachments(self, parsed_email: dict[str, Any]) -> list[ExtractedImage]:
```
Comment (Goldziher) [inline kreuzberg/_extractors/_email.py:140]: This method should be refactored

File: kreuzberg/_extractors/_html.py [hunk 15]
```diff
@@ -32,12 +43,79 @@ def extract_bytes_sync(self, content: bytes) -> ExtractionResult:
...
+                    self._process_images_with_ocr, extraction_result.images
+                )
+
         return self._apply_quality_processing(extraction_result)
 
     def extract_path_sync(self, path: Path) -> ExtractionResult:
         content = path.read_bytes()
         return self.extract_bytes_sync(content)
+
+    def _extract_images_from_html(self, html_content: str) -> list[ExtractedImage]:
```
Comment (Goldziher) [inline kreuzberg/_extractors/_html.py:64]: This method should be refactored

File: kreuzberg/_extractors/_image.py [hunk 16]
```diff
@@ -26,33 +27,17 @@
...
-        "image/jpx": "jpx",
-        "image/jpm": "jpm",
-        "image/mj2": "mj2",
-        "image/webp": "webp",
-        "image/x-portable-anymap": "pnm",
-        "image/x-portable-bitmap": "pbm",
-        "image/x-portable-graymap": "pgm",
-        "image/x-portable-pixmap": "ppm",
-    }
+    IMAGE_MIME_TYPE_EXT_MAP: ClassVar[Mapping[str, str]] = IMAGE_MIME_TO_EXT
```
Comment (Goldziher) [inline kreuzberg/_extractors/_image.py:30]: This was moved into a constant out of this file. Remove the class method and use the constant directly

File: kreuzberg/_extractors/_pandoc.py [hunk 17]
```diff
@@ -570,6 +585,59 @@ def _extract_file_sync(self, path: Path) -> str:
...
+                    "markdown",
+                    "-o",
+                    "/dev/null",
+                ]
+
+                await run_process(cmd)
+
+                if media_dir.exists():
+                    for img_path in media_dir.rglob("*"):
+                        if img_path.is_file() and img_path.suffix.lower() in {
```
Comment (Goldziher) [inline kreuzberg/_extractors/_pandoc.py:574]: Use our shared constants from mime types, or place these as a constant at the top of this file

File: kreuzberg/_extractors/_pdf.py [hunk 18]
```diff
@@ -163,6 +201,89 @@ def _validate_extracted_text(self, text: str, corruption_threshold: float = 0.05
 
         return (len(corruption_matches) / len(text)) < corruption_threshold
 
+    async def _extract_images_from_playa(self, doc: Document) -> list[ExtractedImage]:
+        async def extract_single_image(page_num: int, img_index: int, img_obj: Any) -> ExtractedImage | None:
```
Comment (Goldziher) [inline kreuzberg/_extractors/_pdf.py:203]: Don't online this inside another function. Also, refactor this - use chain and comorehensions

File: kreuzberg/_extractors/_pdf.py [hunk 19]
```diff
@@ -163,6 +201,89 @@ def _validate_extracted_text(self, text: str, corruption_threshold: float = 0.05
...
+                img_counter += 1
+
+        if tasks:
+            results = await asyncio.gather(*tasks)
+            return [img for img in results if img is not None]
+
+        return []
+
+    def _extract_images_from_playa_sync(self, doc: Document) -> list[ExtractedImage]:
+        def extract_single_image(page_num: int, img_index: int, img_obj: Any) -> ExtractedImage | None:
```
Comment (Goldziher) [inline kreuzberg/_extractors/_pdf.py:240]: Don't online inside another function, bad engineering

File: kreuzberg/_extractors/_pdf.py [hunk 20]
```diff
@@ -163,6 +201,89 @@ def _validate_extracted_text(self, text: str, corruption_threshold: float = 0.05
...
+        images = []
+        max_workers = min(8, len(jobs))
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures = {executor.submit(extract_single_image, *job): i for i, job in enumerate(jobs)}
+            for future in as_completed(futures):
+                result = future.result()
+                if result:
+                    images.append(result)
+
+        images.sort(key=lambda x: int((x.filename or "page_0_image_0.jpg").split("_")[-1].split(".")[0]))
```
Comment (Goldziher) [inline kreuzberg/_extractors/_pdf.py:282]: This is ugly, entire function should be cleaned

File: kreuzberg/_extractors/_pdf.py [hunk 21]
```diff
@@ -302,25 +423,51 @@ def _extract_pdf_searchable_text_sync(self, path: Path) -> str:
...
+                            page_height=height,
+                            target_dpi=self.config.target_dpi,
+                            max_dimension=self.config.max_image_dimension,
+                            min_dpi=self.config.min_dpi,
+                            max_dpi=self.config.max_dpi,
+                        )
+                    else:
+                        optimal_dpi = self.config.target_dpi
+
+                    scale = optimal_dpi / 72.0
```
Comment (Goldziher) [inline kreuzberg/_extractors/_pdf.py]: Use a constant fir 72.0, we should replace all places we have it with a single constants from constants.py

File: kreuzberg/_extractors/_presentation.py [hunk 22]
```diff
@@ -106,8 +124,49 @@ def _extract_pptx(self, file_contents: bytes) -> ExtractionResult:
             chunks=[],
         )
 
+        if self.config.extract_images:
+            result.images = self._extract_images_from_pptx(presentation)
+
         return self._apply_quality_processing(result)
 
+    def _extract_images_from_pptx(self, presentation: Presentation) -> list[ExtractedImage]:
```
Comment (Goldziher) [inline kreuzberg/_extractors/_presentation.py:137]: Code needs refactoring and optimizations - nested loops, try blocks, many ifs. Applies to entire file

File: kreuzberg/_ocr/_tesseract.py [hunk 23]
```diff
@@ -1092,6 +1093,55 @@ def _get_file_info(self, path: Path) -> dict[str, Any]:
...
+        run_config = self._prepare_tesseract_run_config(**kwargs)
+        config_dict: dict[str, Any] = {
+            **run_config["remaining_kwargs"],
+            "language": run_config["language"],
+            "psm": run_config["psm"],
+        }
+
+        optimal_workers = get_optimal_worker_count(len(paths), cpu_intensive=True)
+
+        with ProcessPoolExecutor(max_workers=optimal_workers) as pool:
```
Comment (Goldziher) [inline kreuzberg/_ocr/_tesseract.py:1119]: Refactor this function - nested loops and try

File: kreuzberg/_types.py [hunk 24]
```diff
@@ -761,6 +787,39 @@ class ExtractionConfig(ConfigDict):
...
+    """Whether to perform OCR on extracted images."""
+    deduplicate_images: bool = True
+    """Whether to remove duplicate images using CRC32 checksums."""
+    image_ocr_backend: OcrBackendType | None = None
+    """Optional override OCR backend for image OCR; falls back to ocr_backend when None."""
+    image_ocr_min_dimensions: tuple[int, int] = (50, 50)
+    """Minimum (width, height) in pixels for image OCR eligibility."""
+    image_ocr_max_dimensions: tuple[int, int] = (10000, 10000)
+    """Maximum (width, height) in pixels for image OCR eligibility."""
+    image_ocr_formats: frozenset[str] = frozenset(
```
Comment (Goldziher) [inline kreuzberg/_types.py:857]: Use the shared constants from _mime_types.py

File: kreuzberg/_utils/_image_preprocessing.py [hunk 25]
```diff
@@ -178,20 +164,15 @@ def normalize_image_dpi(
         scale_factor *= dimension_scale
         dimension_clamped = True
 
-    # Resize image
     try:
```
Comment (Goldziher) [inline kreuzberg/_utils/_image_preprocessing.py:168]: Nested try blocks. This code needs to be refactored

File: kreuzberg/_utils/_image_preprocessing.py [hunk 26]
```diff
@@ -178,20 +164,15 @@ def normalize_image_dpi(
...
-        # Use LANCZOS for high-quality downscaling, BICUBIC for upscaling
-        # Handle different PIL versions
         try:
-            # Modern PIL version
             if scale_factor < 1.0:
```
Comment (Goldziher) [inline kreuzberg/_utils/_image_preprocessing.py:170]: No need to handle different PIL version. Assume latest always.

File: kreuzberg/_api/main.py [hunk 27]
```diff
@@ -54,7 +58,7 @@ class ImageOCRResultDict(TypedDict):
     """TypedDict for image OCR result JSON representation."""
 
     image: ExtractedImageDict
-    ocr_result: Any  # ExtractionResult serialized
+    ocr_result: Any
```
Comment (Goldziher) [inline kreuzberg/_api/main.py:61]: no Any here, add proper types for this.
